### ISSUE TITLE
Highest Eyetracking Framerate only 60 FPS

### ISSUE THREAD

**Issue by wunan96nj** [microsoft/MixedReality-EyeTracking-Sample#3]
Hi,

I have built and deployed the project to HoloLens. However, from the debug outputs, the highest supportedFrameRates is only 60fps. Are there any other settings I should enable to make it support 90fps framerate?

I have also tried to build a project with unreal, using the codes introduced in the "using native engine" (https://learn.microsoft.com/en-us/windows/mixed-reality/develop/native/extended-eye-tracking-native). However, the highest supported framerate is still 60fps.

I would appreciate it if you could provide me with any information for supporting the 90fps eye tracking on HoloLens 2.

Best.

**Comment by Bowen-MS** [comment]
Hello, could you go to HoloLens Settings -> About, then check about the OS Build number?

**Comment by wunan96nj** [comment]
It is 20348.1535. The Activation status is "Windows is not activated", will it affect the eye-tracking framerate?

**Comment by Bowen-MS** [comment]
No, the activation shouldn't matter. 

We are aware of the issues and working to fix it in later OS updates. 

For now to unblock your side, I suggest install the insider builds following the steps [here](https://learn.microsoft.com/en-us/hololens/hololens-insider#start-receiving-insider-builds). The 90fps works fine in the insider builds.

**Comment by wunan96nj** [comment]
Thanks, it works on the insider builds. By the way, do you also have a code sample working with the native engine?

**Comment by Bowen-MS** [comment]
No, but you can refer to [this page](https://learn.microsoft.com/en-us/windows/mixed-reality/develop/native/extended-eye-tracking-native) for some code snippets used in native engine.

**Comment by wunan96nj** [comment]
Thank you very much for the information!

Another question is, I was previously using the following code for recording gaze data.

    uint64_t previous_time = NDKGetTime();
    while (!bStop)
    {  
        if (NDKGetTime() - previous_time >= 3)
        //if (1)
        {
            UE_LOG(LogTemp, Warning, TEXT("Eyetracker connected"));
            FEyeTrackerGazeData GazeData;
            previous_time = NDKGetTime();
            UE_LOG(LogTemp, Warning, TEXT("Before GetGazeData: %llu"), NDKGetTime());
            UEyeTrackerFunctionLibrary::GetGazeData(GazeData);
            UE_LOG(LogTemp, Warning, TEXT("Get GazeData"));
            FVector Start = GazeData.GazeOrigin;
            FVector End = GazeData.GazeOrigin + GazeData.GazeDirection * 100;

            UE_LOG(LogTemp, Warning, TEXT("After GetGazeData: %llu"), NDKGetTime());
            UE_LOG(LogTemp, Warning, TEXT("Start value: %s"), *Start.ToString());
            UE_LOG(LogTemp, Warning, TEXT("End value: %s"), *End.ToString());


        }
  }

The code was mainly following the [link](https://learn.microsoft.com/en-us/windows/mixed-reality/develop/unreal/unreal-gaze-input).

I set the interval time as 3ms for high fps testing. The debug results show that the time interval between two readings is indeed 3ms, and the gaze values are different in each reading. My question is, do you think the code is performing eye-tracking at 333fps (with an interval of 3ms)?

In other words, the [link ](https://learn.microsoft.com/en-us/windows/mixed-reality/design/eye-tracking) says, "The Eye Tracking API](https://learn.microsoft.com/en-us/uwp/api/windows.perception.people.eyespose) provides the information about what the user is looking at as a single eye-gaze ray (gaze origin and direction) at approximately 30 FPS (30 Hz)." Unreal seems using different API according to the [link](https://learn.microsoft.com/en-us/windows/mixed-reality/develop/unreal/unreal-gaze-input), is it running at a higher fps? Or can I say that HoloLens's eyetracking is running at a much faster fps (>333fps), and the unreal's API allows me to read gaze data at whatever fps according to how frequently I call the API for reading gaze data, while the windows.perception.people.eyespose has already set the time interval (frequency) for reading gaze data?

I would appreciate it If you could help clarify how the APIs work, and suggest me any other ideas for recording gaze data at higher FPS (~90) with a main gaming thread running at ~30 FPS.

Best

**Comment by Bowen-MS** [comment]
The `Windows.Perception.People.EyePose` API provides the 30fps eye tracking, which means the eye tracker on HoloLens produces 30 different eye pose data every second. No matter how frequent you call the API, you always get the value from those 30 data for each second. The API returns the gaze data with closest timestamp from what you requested. You could compare your input timestamp and the returned one associated with eye gaze to see the difference. 

With the extended eye tracking, you could enable 60/90fps eye tracking, which means the eye tracker on HoloLens produces 60/90 different eye pose data every second now, so you could get value from those 60/90 data for each second, i.e. you could get eye pose with higher precision timestamp.

To record the 90fps gaze data, I suggest use the extended eye tracking API `EyeGazeTracker.TryGetReadingAfterTimestamp`. You could record the data every 1 second. 

```
// Record every 1 second
var endTime = DateTimeOffset.Now;
var startTime = endTime - TimeSpan.FromSeconds(1);
var currentTime = startTime;
while (currentTIme < endTime) {
   var reading = _gazeGazeTracker.TryGetReadingAfterTimestamp(currentTime.DateTime);
   currentTime = reading.Timestamp;

   // record the reading data
}
```

**Comment by wunan96nj** [comment]
Thank you very much for the responses! 

Do you know if unreal's Gaze input ([link](https://learn.microsoft.com/en-us/windows/mixed-reality/develop/unreal/unreal-gaze-input)) is using Windows.Perception.People.EyePose API?

**Comment by Bowen-MS** [comment]
As the [release](https://learn.microsoft.com/en-us/hololens/hololens-release-notes#windows-holographic-version-22h2---february-2023-update) of February 2023, the 90fps features is also available in the normal builds starting from `20348.1537`. No need to use insider builds now. 

Update the readme to indicate the minimum OS build number (#4), and resolve this issue.
